provide a class for partitions with bounded length and minimal part

mantepse opened the pull request:
We provide a new class, `PartitionsSmallestGE` to handle subsets of the constraints `length`, `min_length`, `max_length` and `min_part`, to make cardinality computations with such constraints reasonable.

Fixes #38897, #39107

mantepse committed:
provide a class for partitions with bounded length and minimal part

fchapoton reviewed (approved):
ok. Please avoid doing a lot of cleanup work in this kind of ticket.

mantepse committed:
rename new class to Partitions_parts_length_restricted
and generalize it to handle also max_part

mantepse committed:
factor out number_of_partitions_length_max_part and further speedups

mantepse committed:
Merge branch 'sagemath:develop' into partitions/cardinality

mantepse committed:
do not inherit from IntegerListsLex

mantepse committed:
micro optimisations in Partitions.__classcall_private__

mantepse committed:
provide a catch-all class for partitions with unrestricted size, make PartitionsInBox accessible

mantepse committed:
make helper function compute partitions with bounded parts and bounded length, better naming, better classcall_private

mantepse committed:
improve computation of partitions with bound on maximal part

mantepse committed:
Merge branch 'sagemath:develop' into partitions/cardinality

mantepse committed:
Merge branch 'develop' into partitions/cardinality

tscrim commented on src/sage/combinat/partition.py:
```diff
@@ -8784,14 +8930,198 @@ def cardinality(self):
+                and (not mu
+                     or (min(mu) >= self._min_part
+                         and max(mu) <= self._max_part
+                         and self._min_length <= len(mu) <= self._max_length)))
```
```suggestion
        if mu not in _Partitions:
            return False
        if not self._n:
            return not mu
        return (sum(mu) == self._n
                and mu[-1] >= self._min_part
                and mu[0] <= self._max_part
                and self._min_length <= len(mu) <= self._max_length)
```

tscrim commented on src/sage/combinat/partition.py:
```diff
@@ -8784,14 +8930,198 @@ def cardinality(self):
+                                   max_part=self._max_part,
+                                   min_length=self._min_length,
+                                   max_length=self._max_length,
+                                   element_constructor=Partition)
```
```suggestion
        yield from IntegerListsLex(self._n, max_slope=0,
                                   min_part=self._min_part,
                                   max_part=self._max_part,
                                   min_length=self._min_length,
                                   max_length=self._max_length,
                                   element_constructor=lambda x: self.element_class(self, x))
```
Otherwise this is yielding elements in the wrong parent.

mantepse replied:
good catch!

tscrim commented on src/sage/combinat/partition.py:
```diff
@@ -6728,6 +6806,70 @@ def from_core_and_quotient(self, core, quotient):
+            mu = Partition(x)
+        except ValueError:
+            return False
+        return mu in Partitions(mu.size(), **self._restrictions)
```
```suggestion
        try:
            return mu in Partitions(sum(mu), **self._restrictions)
        except Exception:
            return False
```

tscrim commented on src/sage/combinat/partition.py:
```diff
@@ -8784,14 +8930,198 @@ def cardinality(self):
+            sage: Partitions_length_and_parts_restricted(9, 1, 9, 2, 9)
+            Partitions of 9 whose parts are at least 2
+            sage: Partitions_length_and_parts_restricted(9, 3, 5, 2, 9)
+            Partitions of 9 having length between 3 and 5 and whose parts are at least 2
```
I think each type of conditions should be printed at least once (see the Codecov).

mantepse replied:
I agree, done!

tscrim reviewed (changes requested):
This likely would be the last set of changes from me before a positive review.

mantepse committed:
Apply suggestions from code review

in particular, return elements of the correct parent in `Partitions_length_and_parts_restricted`

Co-authored-by: Travis Scrimshaw <clfrngrown@aol.com>

mantepse committed:
fix oversight in previous commit, add some tests, fix typo

mantepse committed:
replace _restricted with _constrained to avoid misconceptions

mantepse committed:
add doctests for containment with trailing zero

mantepse committed:
fix typo

mantepse committed:
fix another typo

mantepse committed:
yet another thinko

mantepse committed:
fix and test __contains__

mantepse committed:
revert to previous behaviour

mantepse committed:
make __contains__ self-contained

mantepse committed:
Merge branch 'sagemath:develop' into partitions/cardinality

mantepse committed:
Merge branch 'sagemath:develop' into partitions/cardinality

fchapoton reviewed (approved):
let's move forward and keep further changes for later
